storage/innobase/log/log0log.cc

```diff
@@ -87,6 +87,117 @@ void log_t::set_capacity()
 }
 
 #ifdef HAVE_PMEM
+# ifdef __x86_64__
+void (*log_t::pmem_persister)(const void *, size_t);
+
+static void log_clflush(const void *buf, size_t size)
+{
+  for (uintptr_t u= uintptr_t(buf) & ~(CPU_LEVEL1_DCACHE_LINESIZE),
+         end= uintptr_t(buf) + size;
+       u < end; u+= CPU_LEVEL1_DCACHE_LINESIZE)
+    __asm__ __volatile__("clflush %0"::
+                         "m"(*reinterpret_cast<const char*>(u)) : "memory");
+}
+
+static void log_clflushopt(const void *buf, size_t size)
+{
+  for (uintptr_t u= uintptr_t(buf) & ~(CPU_LEVEL1_DCACHE_LINESIZE),
+         end= uintptr_t(buf) + size;
+       u < end; u+= CPU_LEVEL1_DCACHE_LINESIZE)
+    __asm__ __volatile__(".byte 0x66; clflush %0" /* clflushopt */::
+                         "m"(*reinterpret_cast<const char*>(u)) : "memory");
+  __asm__ __volatile__("sfence":::"memory");
+}
+
+static void log_clwb(const void *buf, size_t size)
+{
+  for (uintptr_t u= uintptr_t(buf) & ~(CPU_LEVEL1_DCACHE_LINESIZE),
+         end= uintptr_t(buf) + size;
+       u < end; u+= CPU_LEVEL1_DCACHE_LINESIZE)
+    __asm__ __volatile__(".byte 0x66; xsaveopt %0" /* clwb */::
+                         "m"(*reinterpret_cast<const char*>(u)) : "memory");
+  __asm__ __volatile__("sfence":::"memory");
+}
+
+#  include <cpuid.h>
+inline void log_t::pmem_persister_init()
+{
+  uint32_t eax= 0, ebx= 0, ecx= 0, edx= 0;
+  __cpuid_count(7, 0, eax, ebx, ecx, edx);
+  if (ebx & 1U<<24 /* CLWB */)
+    pmem_persister= log_clwb;
+  else if (ebx & 1U<<23 /* CLFLUSHOPT */)
+    pmem_persister= log_clflushopt;
+  else
+    pmem_persister= log_clflush;
```

Comment:
Initially I had `__cpuid(7, eax, ebx, ecx, edx)` here. It would fail to initialize the `ecx` register to 0, and hence all the 4 registers were 0 after executing the `cpuid` instruction. After I fixed that, I got the expected result (`log_clwb`) on our PMEM-equipped system. My development system only implements `clflush`.